### ISSUE TITLE
bring system support in fusell.py to match fuse.py

### ISSUE THREAD

**Issue by terencehonles** [fusepy/fusepy#85]
There have been contributions such as #72 #54 that don't apply patches to fusell.py. We should spend some time making sure the definitions match/are shared.

**Comment by s-m-e** [comment]
This is one of the things I had in mind when I suggested the following (see #80):

> Breaking the files fuse.py and fusell.py into files (sub-modules) specific to operating systems and architectures and common files. (And for people loving single-file modules, if requested, adding a "compiler" script for producing a single-file distribution/release version.)

It's a breaking change in terms of development and, from my point of view, it should happen only after most existing pull requests have either been merged or rejected (if this is done at all). My idea was/is a structure looking somewhat like this:

```
* setup.py
* compile_monolithic_module_file.py
- src
    - fusepy
        * __init__.py
    - fusepyll
        * __init__.py
    - fusepy_lib
        * __init__.py
        * common.py
        - arch
            - linux_x86
                * __init__.py
                [...]
            - linux_x86_64
            - linux_mips
            - linux_ppc
            - linux_ppc64
            - bsd_x86
            - bsd_x86_64
            - osx_x86_64
            - osx_macfuse_x86_64
            - win_x86
            - win_x86_64
[...]
```

The file `src/fusepy_lib/__init__.py` could for instance contain some clever logic based on [importlib](https://docs.python.org/3/library/importlib.html), automatically importing appropriate arch sub-module. The file `common.py` could contain all infrastructure common to `fusepy` and `fusepyll`, which is not specific to architectures or operating systems.

For use cases where a single monolithic module file makes sense, it has already been [requested](https://github.com/fusepy/fusepy/issues/75#issuecomment-358754801) to add some sort of a compiler script, which should not be too complicated - so this is what a script like `compile_monolithic_module_file.py` would do.

A structure of this kind makes life a lot easier when maintaining or extending the module ... thoughts?

**Comment by eestrada** [comment]
Would `compile_monolithic_module_file.py` be called in `setup.py`? That makes most sense to me, but maybe that is overkill, since `setup.py` and the pypi package can just have all the files come along for the install, which would make it easier for developers/users to debug and create issues and merge requests (since what is installed more or less matches what the repo shows).

**Comment by s-m-e** [comment]
I guess it depends on the actual use case (@itdaniher ?). I'd assume that people who need a monolithic module file do not use `pip` or `setup.py` in the environment they deploy into. So I'd again guess that a script like `compile_monolithic_module_file.py` should only be offered for development purposes in the repo but not distributed in PyPI packages. On the other hand, it could be distributed as a small CLI tool as part of the released packages but not automatically triggered by `setup.py`.

**Comment by terencehonles** [comment]
@s-m-e the layout seems fine, but possibly a bit overkill with tons of small files (we'll have to see). @eestrada I imagine the compile script is just there for those who want to use the script w/o `setup.py` however if we're going to have a script that compiles the module into one file we're going to want to include that as part of the test suite so the compile script doesn't end up generating a broken/incomplete package.

**Comment by terencehonles** [comment]
@s-m-e looking at #90 and how fusell isn't already exported we can probably do something like

```
src/
└── fuse
    ├── common.py
    └── __init__.py
```

and have `__init__.py` export both what `fuse.py` and `fusell.py` export, and have the compile script flatten the files to the "original" interface

**Comment by s-m-e** [comment]
@terencehonles @eestrada Yep, the compile script and its products must be tested just like the regular package, agreed. 

The `overkill` structure with many small files looks a bit scary, I'd agree. On the other hand, please keep in mind the following: This Python package actually is a "distribution" of two packages on equal footing: `fuse` and `fusell`. People do not run, say, `import fusell from big_package` or `import fuse from big_package`. They run either `import fuse` or `import fusell`. I was considering to keep things this way for backwards compatibility ... Those two "independent" package share a lot of common infrastructure, so this is where a third top-level "library package" (`fusepy_lib` in my example) comes into play from my point of view. Up to this point, we're talking about 3 files instead of 2. Now the architectures and operating systems kick in. Across all of them, the way the actual FUSE library (`libfuse`) is accessed vary. Structures vary. So instead of having tons of cryptic `if`-`elif`-`else` blocks (like in the current version of `fuse` and `fusell`), I am suggesting to keep things clean: One file per architecture/operating system combination, containing nothing but library import (`find_library` and `load_library`) and (structure) type definitions. Debugging, adding support for and dropping support for a specific architecture/operating system combination becomes a lot easier this way.

**Comment by terencehonles** [comment]
Agreed, however you could expose `fusell` as a shell which just imports what it is going to re-export from the `fuse` package, and potentially put deprecation warnings that eventually that shell might go away and the preferred way to import the package would be through `fuse`.

Anyways, I'll take a stab at it and create a PR for splitting things.

**Comment by s-m-e** [comment]
@terencehonles If I am understanding your idea correctly, I very much like it. You are proposing to create what I dubbed a `big_package` while the files `fuse.py` and `fusell.py` become stubs, marked as deprecated, which only import their functionality from the `big_package`? I'd vote for `fusepy` as the name for the big package :)

**Comment by terencehonles** [comment]
I was thinking `fuse/`, but `fusepy/` could work
